jackson-core/src/main/java/io/micronaut/jackson/core/tree/MicronautTreeCodec.java

```diff
@@ -0,0 +1,134 @@
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.*;
+
+public final class MicronautTreeCodec {
+    private static final MicronautTreeCodec INSTANCE = new MicronautTreeCodec(JsonConfig.DEFAULT);
+
+    private final JsonConfig config;
+
+    private MicronautTreeCodec(JsonConfig config) {
+        this.config = config;
+    }
+
+    public static MicronautTreeCodec getInstance() {
+        return INSTANCE;
+    }
+
+    public MicronautTreeCodec withConfig(JsonConfig config) {
+        return new MicronautTreeCodec(config);
+    }
+
+    public JsonNode readTree(JsonParser p) throws IOException {
+        switch (p.hasCurrentToken() ? p.currentToken() : p.nextToken()) {
+            case START_OBJECT:
+                Map<String, JsonNode> map = new LinkedHashMap<>();
+                while (p.nextToken() != JsonToken.END_OBJECT) {
+                    String name = p.currentName();
+                    p.nextToken();
+                    map.put(name, readTree(p));
+                }
+                return JsonNode.createObjectNode(map);
+            case START_ARRAY:
+                List<JsonNode> list = new ArrayList<>();
+                while (p.nextToken() != JsonToken.END_ARRAY) {
+                    list.add(readTree(p));
+                }
+                return JsonNode.createArrayNode(list);
+            case VALUE_STRING:
+                return JsonNode.createStringNode(p.getText());
+            case VALUE_NUMBER_INT:
+                if (config.useBigIntegerForInts()) {
+                    return JsonNode.createNumberNode(p.getBigIntegerValue());
+                } else {
+                    // technically, we could get an unsupported number value here.
+                    return JsonNode.createNumberNodeImpl(p.getNumberValue());
+                }
+            case VALUE_NUMBER_FLOAT:
+                if (config.useBigDecimalForFloats()) {
+                    return JsonNode.createNumberNode(p.getDecimalValue());
+                } else {
+                    // technically, we could get an unsupported number value here.
+                    return JsonNode.createNumberNodeImpl(p.getNumberValue());
+                }
+            case VALUE_TRUE:
+                return JsonNode.createBooleanNode(true);
+            case VALUE_FALSE:
+                return JsonNode.createBooleanNode(false);
+            case VALUE_NULL:
+                return JsonNode.nullNode();
+            default:
+                throw new UnsupportedOperationException("Unsupported token: " + p.currentToken());
+        }
+    }
+
+    public void writeTree(JsonGenerator generator, JsonNode tree) throws IOException, JsonProcessingException {
+        if (tree.isObject()) {
+            generator.writeStartObject();
+            for (Map.Entry<String, JsonNode> entry : tree.entries()) {
+                generator.writeFieldName(entry.getKey());
+                writeTree(generator, tree);
```

Comment:
This recursion seems unusual, what breaks this loop?

Reply:
Yes, this is a bug from copy-pasting. It's concerning that no tests caught this, seems like they didn't transfer over. I will need to check the test coverage of the PR.